File: Low Level Design (LLD)/Problems/07. SocialApp/Cpp/src/SocialMediaApp.cpp
#include "SocialMediaApp.h"

std::shared_ptr<User> SocialMediaApp::createUser(const std::string& name) {
    auto user = std::make_shared<User>(userId_++, name);
    users_[user->getId()] = user;
    return user;
}

std::shared_ptr<Post> SocialMediaApp::createPost(std::shared_ptr<User> author, const std::string& content) {
    auto post = std::make_shared<Post>(postId_++, author, content);
    posts_[post->getId()] = post;
    author->addPost(post);
    return post;
}

std::shared_ptr<Comment> SocialMediaApp::createComment(std::shared_ptr<User> author, const std::string& content) {
    auto comment = std::make_shared<Comment>(commentId_++, author, content);
    comments_[comment->getId()] = comment;
    return comment;
}

std::shared_ptr<Story> SocialMediaApp::createStory(std::shared_ptr<User> author, const std::string& content) {
    auto story = std::make_shared<Story>(storyId_++, author, content);
    stories_[story->getId()] = story;
    return story;
}

std::shared_ptr<User> SocialMediaApp::getUser(int id) {
    return users_.find(id) != users_.end() ? users_[id] : nullptr;
}
std::shared_ptr<Post> SocialMediaApp::getPost(int id) {
    return posts_.find(id) != posts_.end() ? posts_[id] : nullptr;
}
std::shared_ptr<Comment> SocialMediaApp::getComment(int id) {
    return comments_.find(id) != comments_.end() ? comments_[id] : nullptr;
}
std::shared_ptr<Story> SocialMediaApp::getStory(int id) {
    return stories_.find(id) != stories_.end() ? stories_[id] : nullptr;
}

File: Low Level Design (LLD)/Problems/07. SocialApp/Java/services/SocialMediaApp.java
package services;

import models.*;

import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;

public class SocialMediaApp {
    private final AtomicInteger userId = new AtomicInteger(0);
    private final AtomicInteger postId = new AtomicInteger(0);
    private final AtomicInteger commentId = new AtomicInteger(0);
    private final AtomicInteger storyId = new AtomicInteger(0);

    private final Map<Integer, User> users = new HashMap<>();
    private final Map<Integer, Post> posts = new HashMap<>();
    private final Map<Integer, Comment> comments = new HashMap<>();
    private final Map<Integer, Story> stories = new HashMap<>();

    public User createUser(String name) {
        int id = userId.getAndIncrement();
        User user = new User(id, name);
        users.put(id, user);
        return user;
    }

    public Post createPost(User author, String content) {
        int id = postId.getAndIncrement();
        Post post = new Post(id, author, content);
        posts.put(id, post);
        author.addPost(post);
        return post;
    }

    public Comment createComment(User author, String content) {
        int id = commentId.getAndIncrement();
        Comment comment = new Comment(id, author, content);
        comments.put(id, comment);
        return comment;
    }

    public Story createStory(User author, String content) {
        int id = storyId.getAndIncrement();
        Story story = new Story(id, author, content);
        stories.put(id, story);
        return story;
    }

    // Getters by id
    public Optional<User> getUser(int id) {
        return Optional.ofNullable(users.get(id));
    }

    public Optional<Post> getPost(int id) {
        return Optional.ofNullable(posts.get(id));
    }

    public Optional<Comment> getComment(int id) {
        return Optional.ofNullable(comments.get(id));
    }

    public Optional<Story> getStory(int id) {
        return Optional.ofNullable(stories.get(id));
    }
}
